In [1]:
from sklearn.datasets import load_diabetes
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
diabetes = load_diabetes()

diabetes : 442 명의 당뇨병 환자 각각에 대해 10 개의 기준 변수, 연령, 성별, 체질량 지수, 평균 혈압 및 6 개의 혈청 측정 값

In [2]:
diabetes

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990842, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06832974, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286377, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04687948,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452837, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00421986,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

In [3]:
print('data :', diabetes['data'])

data : [[ 0.03807591  0.05068012  0.06169621 ... -0.00259226  0.01990842
  -0.01764613]
 [-0.00188202 -0.04464164 -0.05147406 ... -0.03949338 -0.06832974
  -0.09220405]
 [ 0.08529891  0.05068012  0.04445121 ... -0.00259226  0.00286377
  -0.02593034]
 ...
 [ 0.04170844  0.05068012 -0.01590626 ... -0.01107952 -0.04687948
   0.01549073]
 [-0.04547248 -0.04464164  0.03906215 ...  0.02655962  0.04452837
  -0.02593034]
 [-0.04547248 -0.04464164 -0.0730303  ... -0.03949338 -0.00421986
   0.00306441]]


In [31]:
# 변수 명 확인
print('feature names :', diabetes['feature_names'])

feature names : ['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']


In [5]:
print('target :', diabetes['target'])

target : [151.  75. 141. 206. 135.  97. 138.  63. 110. 310. 101.  69. 179. 185.
 118. 171. 166. 144.  97. 168.  68.  49.  68. 245. 184. 202. 137.  85.
 131. 283. 129.  59. 341.  87.  65. 102. 265. 276. 252.  90. 100.  55.
  61.  92. 259.  53. 190. 142.  75. 142. 155. 225.  59. 104. 182. 128.
  52.  37. 170. 170.  61. 144.  52. 128.  71. 163. 150.  97. 160. 178.
  48. 270. 202. 111.  85.  42. 170. 200. 252. 113. 143.  51.  52. 210.
  65. 141.  55. 134.  42. 111.  98. 164.  48.  96.  90. 162. 150. 279.
  92.  83. 128. 102. 302. 198.  95.  53. 134. 144. 232.  81. 104.  59.
 246. 297. 258. 229. 275. 281. 179. 200. 200. 173. 180.  84. 121. 161.
  99. 109. 115. 268. 274. 158. 107.  83. 103. 272.  85. 280. 336. 281.
 118. 317. 235.  60. 174. 259. 178. 128.  96. 126. 288.  88. 292.  71.
 197. 186.  25.  84.  96. 195.  53. 217. 172. 131. 214.  59.  70. 220.
 268. 152.  47.  74. 295. 101. 151. 127. 237. 225.  81. 151. 107.  64.
 138. 185. 265. 101. 137. 143. 141.  79. 292. 178.  91. 116.  86. 12

In [38]:
# 훈련 데이터와 테스트 데이터 분리
from sklearn.model_selection import train_test_split
X_train, X_test, y_train,y_test = train_test_split(
diabetes['data'],diabetes['target'], random_state = 42)

In [39]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(331, 10)
(331,)
(111, 10)
(111,)


In [40]:
df_diabetes = pd.DataFrame(X_train, columns = diabetes['feature_names'])
df_diabetes.head()

age       sex       bmi        bp        s1        s2        s3  \
0 -0.005515 -0.044642  0.042296  0.049415  0.024574 -0.023861  0.074412   
1  0.063504 -0.044642 -0.050396  0.107944  0.031454  0.019354 -0.017629   
2  0.009016 -0.044642  0.055229 -0.005671  0.057597  0.044719 -0.002903   
3  0.016281  0.050680  0.014272  0.001215  0.001183 -0.021355 -0.032356   
4 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   

         s4        s5        s6  
0 -0.039493  0.052280  0.027917  
1  0.023608  0.058039  0.040343  
2  0.023239  0.055684  0.106617  
3  0.034309  0.074968  0.040343  
4  0.034309  0.022692 -0.009362

### k-최근접 이웃 회귀

In [41]:
from sklearn.neighbors import KNeighborsRegressor

In [42]:
# 회귀모델 생성
# distance를 사용할 경우 가중평균 사용
# 기본값은 단순 평균인 uniform
regressor = KNeighborsRegressor(n_neighbors = 3, weights = 'distance')

In [43]:
# 모델 훈련
regressor.fit(X_train,y_train)

KNeighborsRegressor(n_neighbors=3, weights='distance')

In [44]:
print("test 결과 예측 :" , regressor.predict(X_test))

test 결과 예측 : [139.76088773 191.95610684 145.75310919 239.21106285 157.32645636
 130.82244551 242.83086999 174.47190435  98.72692894  80.10871328
  69.91784191 152.82112732 118.8786927  205.80323396  67.4918476
 135.24123226 275.65285103 259.58550189 209.60329883 253.89697324
 143.77421245  72.97899458  93.83001528 182.27411892 132.61776841
 144.22639668 175.30470213 169.13554746  74.98914454 112.4792323
 164.17655707 142.47746626  83.20381379 194.15773797 144.34504253
 245.83928958  72.30808504 158.73419657 178.95420935  92.84996938
  84.21481975 106.34074384 141.212387   160.29892963 211.69745673
  77.45282107  61.91220304  70.40036598 109.98469212 133.1137846
 158.48605702 110.64254832 141.73930777 126.03501836 175.54276425
 143.23588546  68.53735433 167.84144402 151.2170479   74.29613594
 144.79607018 107.52845129  65.08780501  74.52542118 141.44316072
 164.31821128 208.67360288 166.58521049 156.59682433 111.15890821
 106.97150145 195.54319901 187.92059782  72.30111494 121.79331326


In [45]:
# 최근접 이웃 회귀는 보통 분류에 사용하므로 선형회귀 데이터셋에서는 r^2가 0.13으로 낮게 나옴
print("test R^2 : {:.2f}".format(regressor.score(X_test,y_test)))

test R^2 : 0.39


### 선형 회귀

In [46]:
from sklearn.linear_model import LinearRegression

In [47]:
# 선형회귀 모델 훈련
#fit_intercept=True, : y절편을 계산할건지 여부
# normalize=True,  : 정규화를 시킬것인지
# n_jobs=None : 모델 계산에 사용할 작업 수

lr = LinearRegression(normalize = True).fit(X_train,y_train)

In [48]:
# 회귀계수 계산값
print('lr.coef_:', lr.coef_) # 훈련 데이터에서 이끌어낸 속성은 끝에 _를 붙임

# y절편 계산값
print('lr.intercept_:', lr.intercept_)

lr.coef_: [  47.74657117 -241.99180361  531.96856896  381.56529922 -918.49020552
  508.25147385  116.94040498  269.48508571  695.8062205    26.32343144]
lr.intercept_: 151.66516982689936


In [49]:
# 모델 정확도 계산 
print('train score: {:.2f}'.format(lr.score(X_train,y_train)))
print('test score : {:.2f}'.format(lr.score(X_test,y_test)))

train score: 0.52
test score : 0.48


In [52]:
# 테스트의 예측값 생성
prediction = lr.predict(X_test)
print(prediction)

[137.94979889 182.53621462 129.85554049 292.55738727 124.86559124
  89.73987275 255.96040544 177.64691719  87.88678516 107.93733249
  93.26418526 171.11859131  56.06450108 207.22092723  99.79206335
 131.71477524 215.86564416 252.18825848 201.56370519 220.38853067
 204.35710556  88.90708813  68.53068596 190.02202826 153.9129781
 164.02499711 192.83524491 187.16510251  46.60505083 109.31176844
 175.05178809  87.8091125  130.37787111 186.56591449 172.474085
 188.61455112 124.0734688  119.16492145 147.74614571  59.27160039
  70.90301607 109.30498566 170.97369589 156.56765896 168.68372176
  61.54562801  71.19297293 111.94403965  54.31821851 165.81456881
 152.78238871  63.80826526 110.07732165 108.6406539  179.16116053
 156.79589523  93.28778814 212.92031906 119.28369688  70.70909587
 186.68094494 205.65924203 140.02745143 106.54570273 131.56044364
 204.17886366 173.28771583 168.65292359 123.58530603 145.48887871
 182.55324099 200.92532802 234.1265505  148.65425454  82.2675761
 162.10939001 

In [55]:
# mean squared error 계산
from sklearn.metrics import mean_squared_error

print('mse : {:.2f}'.format(mean_squared_error(X_test[:,0],prediction)))

mse : 23793.19


mse가 23793으로 예상보다 크게나옴.

변수의 개수를 줄여볼 것

plot 그리기